// @mantine/core/cjs/components/Menu/MenuContextMenu/MenuContextMenu.cjs
"use client";
const require_get_single_element_child = require("../../../core/utils/get-single-element-child/get-single-element-child.cjs");
const require_use_props = require("../../../core/MantineProvider/use-props/use-props.cjs");
const require_Popover_context = require("../../Popover/Popover.context.cjs");
const require_create_context_menu_handlers = require("../../../utils/Floating/create-context-menu-handlers.cjs");
const require_Menu_context = require("../Menu.context.cjs");
let react = require("react");
//#region packages/@mantine/core/src/components/Menu/MenuContextMenu/MenuContextMenu.tsx
function MenuContextMenu(props) {
	const { children, disabled } = require_use_props.useProps("MenuContextMenu", null, props);
	const child = require_get_single_element_child.getSingleElementChild(children);
	if (!child) throw new Error("Menu.ContextMenu component children should be an element or a component that accepts ref. Fragments, strings, numbers and other primitive values are not supported");
	const ctx = require_Menu_context.useMenuContext();
	const popoverCtx = require_Popover_context.usePopoverContext();
	return (0, react.cloneElement)(child, require_create_context_menu_handlers.createContextMenuHandlers({
		childProps: child.props,
		disabled: disabled || popoverCtx.disabled,
		opened: ctx.opened,
		setReference: popoverCtx.reference,
		open: () => ctx.openDropdown()
	}));
}
MenuContextMenu.displayName = "@mantine/core/MenuContextMenu";
//#endregion
exports.MenuContextMenu = MenuContextMenu;
