// @mantine/core/cjs/components/Menu/MenuSelectableItem/MenuSelectableItem.cjs
"use client";
const require_create_scoped_keydown_handler = require("../../../core/utils/create-scoped-keydown-handler/create-scoped-keydown-handler.cjs");
const require_create_event_handler = require("../../../core/utils/create-event-handler/create-event-handler.cjs");
const require_parse_theme_color = require("../../../core/MantineProvider/color-functions/parse-theme-color/parse-theme-color.cjs");
const require_MantineThemeProvider = require("../../../core/MantineProvider/MantineThemeProvider/MantineThemeProvider.cjs");
const require_DirectionProvider = require("../../../core/DirectionProvider/DirectionProvider.cjs");
const require_UnstyledButton = require("../../UnstyledButton/UnstyledButton.cjs");
const require_Menu_context = require("../Menu.context.cjs");
const require_MenuSub_context = require("../MenuSub/MenuSub.context.cjs");
let react = require("react");
let _mantine_hooks = require("@mantine/hooks");
let react_jsx_runtime = require("react/jsx-runtime");
//#region packages/@mantine/core/src/components/Menu/MenuSelectableItem/MenuSelectableItem.tsx
function MenuSelectableItem({ role, checked, indicator, onSelect, color, closeMenuOnClick, rightSection, children, disabled, dataDisabled, className, style, styles, classNames, buttonRef, others }) {
	const ctx = require_Menu_context.useMenuContext();
	const subCtx = (0, react.use)(require_MenuSub_context.SubMenuContext);
	const theme = require_MantineThemeProvider.useMantineTheme();
	const { dir } = require_DirectionProvider.useDirection();
	const itemRef = (0, react.useRef)(null);
	const handleClick = require_create_event_handler.createEventHandler(others.onClick, () => {
		if (dataDisabled) return;
		onSelect();
		if (closeMenuOnClick) ctx.closeDropdownImmediately();
	});
	const handleMouseMove = require_create_event_handler.createEventHandler(others.onMouseMove, () => {
		if (!ctx.hasSearch) return;
		const dropdown = itemRef.current?.closest("[data-menu-dropdown]");
		if (!dropdown) return;
		dropdown.querySelectorAll("[data-menu-active]").forEach((node) => {
			if (node !== itemRef.current && node.closest("[data-menu-dropdown]") === dropdown) node.removeAttribute("data-menu-active");
		});
	});
	const handleKeydown = require_create_event_handler.createEventHandler(others.onKeyDown, (event) => {
		if (event.key === "ArrowLeft" && subCtx) {
			subCtx.close();
			subCtx.focusParentItem();
		}
	});
	const colors = color ? theme.variantColorResolver({
		color,
		theme,
		variant: "light"
	}) : void 0;
	const parsedThemeColor = color ? require_parse_theme_color.parseThemeColor({
		color,
		theme
	}) : null;
	const renderIndicator = ctx.alignItemsLabels !== "none" || checked;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_UnstyledButton.UnstyledButton, {
		onMouseDown: (event) => event.preventDefault(),
		...others,
		unstyled: ctx.unstyled,
		tabIndex: ctx.menuItemTabIndex,
		...ctx.getStyles("item", {
			className,
			style,
			styles,
			classNames
		}),
		ref: (0, _mantine_hooks.useMergedRef)(itemRef, buttonRef),
		role,
		"aria-checked": checked,
		disabled,
		"data-menu-item": true,
		"data-checked": checked || void 0,
		"data-disabled": disabled || dataDisabled || void 0,
		"data-mantine-stop-propagation": true,
		onClick: handleClick,
		onMouseMove: handleMouseMove,
		onKeyDown: require_create_scoped_keydown_handler.createScopedKeydownHandler({
			siblingSelector: "[data-menu-item]:not([data-disabled])",
			parentSelector: "[data-menu-dropdown]",
			activateOnFocus: false,
			loop: ctx.loop,
			dir,
			orientation: "vertical",
			onKeyDown: handleKeydown
		}),
		__vars: {
			"--menu-item-color": parsedThemeColor?.isThemeColor && parsedThemeColor?.shade === void 0 ? `var(--mantine-color-${parsedThemeColor.color}-6)` : colors?.color,
			"--menu-item-hover": colors?.hover
		},
		children: [
			renderIndicator && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				...ctx.getStyles("itemIndicator", {
					styles,
					classNames
				}),
				"data-checked": checked || void 0,
				children: checked ? indicator : null
			}),
			children && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				...ctx.getStyles("itemLabel", {
					styles,
					classNames
				}),
				"data-menu-item-label": true,
				children
			}),
			rightSection && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				...ctx.getStyles("itemSection", {
					styles,
					classNames
				}),
				"data-position": "right",
				children: rightSection
			})
		]
	});
}
//#endregion
exports.MenuSelectableItem = MenuSelectableItem;
